### PR TITLE
feat: Add basic logger (#20)

### PR DESCRIPTION
**Pull request by yokuze**
Note: both request / access logs: https://github.com/silvermine/lambda-express/issues/28 and sampling have been broken out into a separate ticket: https://github.com/silvermine/lambda-express/issues/26 to be addressed in a different PR.

### CODE REVIEW

File: src/logging/logging-types.ts [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    * method if the log level is set to `trace`.
+    *
+    * @param msg The message to log
+    * @param data Custom data to add to the log message. Added to the log object under the
+    * `data` key.
+    */
+   trace(msg: string, data: unknown): void;
+
+   /**
+   * Log a message at the `debug` log level. The logger will only log a message for this
```
Comment (jthomerson) [inline src/logging/logging-types.ts]: indentation

Comment (jthomerson) [reply src/logging/logging-types.ts]: a bunch more of these in this file

File: src/logging/ConsoleLogger.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+      }
+   }
+
+   /**
+    * @returns `true` if the given level should be logged at this logger's current log
+    * level setting.
+    */
+   protected _shouldLog(level: LogLevel): boolean {
+      // Log if the level is higher priority than the current log level setting.
+      // e.g.  error (50)  >= info (30)
```
Comment (jthomerson) [inline src/logging/ConsoleLogger.ts]: some double spaces here

File: src/logging/is-debug-level-or-more-verbose.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+/**
+ * @returns `true` if the given log level is `'debug'` or a more verbose level (e.g.
+ * `'trace'`).
+ */
+function isDebugLevelOrMoreVerbose(level: LogLevel): boolean {
+   // More verbose levels have lower priority numbers
+   return levels[level] <= levels.debug;
+}
+
+export default isDebugLevelOrMoreVerbose;
```
Comment (jthomerson) [inline src/logging/is-debug-level-or-more-verbose.ts]: any reason why `export default` wasn't just added to the function declaration above?

**Review by jthomerson — CHANGES_REQUESTED** [review]

**Comment by jthomerson** [comment]
Pushed these changes (ignoring all space since there were a lot of whitespace fixes):

```patch
jrthomer@lappy ~/git-clones/silvermine/lambda-express
[ logging_20 ] $ git diff --ignore-all-space ecdccc9..3bf98ba
diff --git a/src/logging/is-debug-level-or-more-verbose.ts b/src/logging/is-debug-level-or-more-verbose.ts
index 430da74..9c0c20e 100644
--- a/src/logging/is-debug-level-or-more-verbose.ts
+++ b/src/logging/is-debug-level-or-more-verbose.ts
@@ -5,9 +5,7 @@ import levels from './levels';
  * @returns `true` if the given log level is `'debug'` or a more verbose level (e.g.
  * `'trace'`).
  */
-function isDebugLevelOrMoreVerbose(level: LogLevel): boolean {
+export default function isDebugLevelOrMoreVerbose(level: LogLevel): boolean {
    // More verbose levels have lower priority numbers
    return levels[level] <= levels.debug;
 }
-
-export default isDebugLevelOrMoreVerbose;
diff --git a/src/logging/logging-types.ts b/src/logging/logging-types.ts
index d52cbc8..e094b67 100644
--- a/src/logging/logging-types.ts
+++ b/src/logging/logging-types.ts
@@ -127,7 +127,7 @@ export type LogLevels = {
 };
 
 /**
- * Configuration to initialize a Logger with.
+ * Configuration to initialize a `Logger` with.
  */
 export interface LoggerConfig {
 
@@ -193,7 +193,7 @@ export interface DebugLogObject extends LogObject {
    remaining: number;
 
    /**
-    * The Lambda event source type that triggered the current Request.
+    * The Lambda event source type that triggered the current `Request`.
     */
    int: LambdaEventSourceType;

```

File: src/logging/logging-types.ts (hunk 1 above)
Comment (yokuze) [reply src/logging/logging-types.ts]: 🤦‍♂️

File: src/logging/is-debug-level-or-more-verbose.ts (hunk 3 above)
Comment (yokuze) [reply src/logging/is-debug-level-or-more-verbose.ts]: Not that I can think of.

**Comment by yokuze** [comment]
@jthomerson Here's the diff of the actual changes:

```diff
diff --git a/src/Router.ts b/src/Router.ts
index c6c1619..bd32fc8 100644
--- a/src/Router.ts
+++ b/src/Router.ts
@@ -16,6 +16,7 @@ import { RouteMatchingProcessorChain } from './chains/RouteMatchingProcessorChai
 import { MatchAllRequestsProcessorChain } from './chains/MatchAllRequestsProcessorChain';
 import { SubRouterProcessorChain } from './chains/SubRouterProcessorChain';
 import Route from './Route';
+import { Optional } from './utils/common-types';

 const DEFAULT_OPTS: RouterOptions = {
    caseSensitive: false,
@@ -27,7 +28,7 @@ export default class Router implements IRouter {
    public readonly routerOptions: RouterOptions;
    private readonly _processors: IRequestMatchingProcessorChain[] = [];

-   public constructor(options?: RouterOptions) {
+   public constructor(options?: Optional<RouterOptions>) {
       this.routerOptions = _.defaults(options, DEFAULT_OPTS);
    }

diff --git a/src/utils/common-types.ts b/src/utils/common-types.ts
index 4f9c984..ed95855 100644
--- a/src/utils/common-types.ts
+++ b/src/utils/common-types.ts
@@ -6,6 +6,8 @@ export interface StringArrayOfStringsMap { [s: string]: string[] }
 export interface KeyValueStringObject { [k: string]: (string | string[] | KeyValueStringObject) }
 // Removes `readonly` modifier and make all keys writable again
 export type Writable<T> = { -readonly [P in keyof T]-?: T[P] };
+// Makes all keys on an object optional
+export type Optional<T> = { [P in keyof T]? : T[P] };
```

**Comment by yokuze** [comment]
@jthomerson if you accept the approach above, I'll open a ticket / PR to add `Optional<T>` to our [toolbox project](https://github.com/silvermine/toolbox/).

**Comment by jthomerson** [comment]
I just pushed a change to use `Partial` instead of our own `Optional`, ADIRL

```patch
jrthomer@lappy ~/git-clones/silvermine/lambda-express
[ logging_20 ] $ git diff
diff --git a/src/Router.ts b/src/Router.ts
index bd32fc8..8bf9848 100644
--- a/src/Router.ts
+++ b/src/Router.ts
@@ -16,7 +16,6 @@ import { RouteMatchingProcessorChain } from './chains/RouteMatchingProcessorChai
 import { MatchAllRequestsProcessorChain } from './chains/MatchAllRequestsProcessorChain';
 import { SubRouterProcessorChain } from './chains/SubRouterProcessorChain';
 import Route from './Route';
-import { Optional } from './utils/common-types';
 
 const DEFAULT_OPTS: RouterOptions = {
    caseSensitive: false,
@@ -28,7 +27,7 @@ export default class Router implements IRouter {
    public readonly routerOptions: RouterOptions;
    private readonly _processors: IRequestMatchingProcessorChain[] = [];
 
-   public constructor(options?: Optional<RouterOptions>) {
+   public constructor(options?: Partial<RouterOptions>) {
       this.routerOptions = _.defaults(options, DEFAULT_OPTS);
    }
 
diff --git a/src/utils/common-types.ts b/src/utils/common-types.ts
index ed95855..4f9c984 100644
--- a/src/utils/common-types.ts
+++ b/src/utils/common-types.ts
@@ -6,8 +6,6 @@ export interface StringArrayOfStringsMap { [s: string]: string[] }
 export interface KeyValueStringObject { [k: string]: (string | string[] | KeyValueStringObject) }
 // Removes `readonly` modifier and make all keys writable again
 export type Writable<T> = { -readonly [P in keyof T]-?: T[P] };
-// Makes all keys on an object optional
-export type Optional<T> = { [P in keyof T]? : T[P] };
 
 export function isStringMap(o: any): o is StringMap {
    if (!_.isObject(o) || _.isArray(o)) { // because arrays are objects
```

**Review by jthomerson — APPROVED** [review]

**Pull request merged**